book/testing/examples/correct/rate_limiter_fixed/rate_limiter.ml

```diff
@@ -0,0 +1,32 @@
+open! Core
+
+type t =
+  { events : Time_ns.t Queue.t
+  ; mutable now : Time_ns.t
+  ; period : Time_ns.Span.t
+  ; rate : int
+  }
+
+let create ~now ~period ~rate =
+  let events = Queue.create () in
+  { events; now; period; rate }
+
+[@@@part "1"];;
+let rec drain_old_events t =
+  match Queue.peek t.events with
+  | None -> ()
+  | Some time ->
+    if Time_ns.Span.( >= ) (Time_ns.diff t.now time) t.period
```

Comment:
I don't like the proposed style because it opens up Span and it's unclear what identifiers might be shadowed in such expressions.

Really, we need modular implicits to make this kind of code nice! (And Flambda, to make it fast again.)